File: 1915.cpp
#include <iostream>
#include <algorithm>
using namespace std;

constexpr int maxLen = 1001;

int dp[maxLen][maxLen];
char map[maxLen][maxLen];
int numRow, numCol;

int main() {
    cin >> numRow >> numCol;
    int num;
    int maxNum = 0;
    
    for(int i = 1; i <= numRow; i++) {
        for(int j = 1; j <= numCol; j++) {
            cin >> map[i][j];
            map[i][j] -= '0';
            dp[i][j] = map[i][j];
            if(map[i][j]) maxNum = 1;
        }
    }
    
    
    for(int i = 1; i <= numRow; i++) {
        for(int j = 1; j <= numCol; j++) {
            if(dp[i][j] && dp[i - 1][j - 1] && dp[i - 1][j] && dp[i][j - 1]) {
                dp[i][j] = min({dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]}) + 1;
                maxNum = max(dp[i][j], maxNum);
            }
        }
    }
    
    
    
    cout << maxNum * maxNum;
}

/*
 011100
 011111
 011111
 011111
 
 011100
 012211
 012322
 012333
*/

File: 1915.py
import sys

input = lambda: sys.stdin.readline().strip()

num_row, num_col = map(int, input().split())

board = [[0] * (num_col + 1) for i in range(num_row + 1)]
dp = [[0] * (num_col + 1) for i in range(num_row + 1)]

ans = 0

for i in range(1, num_row + 1):
    line = list(input())
    
    for j in range(1, num_col + 1):
        dp[i][j] = int(line[j - 1])
        if dp[i][j]:
            ans = 1
    
for i in range(1, num_row + 1):
    for j in range(1, num_col + 1):
        if dp[i][j] and dp[i - 1][j] and dp[i][j - 1] and dp[i - 1][j - 1]:
            dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
            ans = max(ans, dp[i][j])

print(ans * ans)